lib/compiler/aro_translate_c.zig

```diff
@@ -496,16 +675,152 @@ fn transType(c: *Context, scope: *Scope, raw_ty: Type, source_loc: TokenIndex) T
         .long_double => return ZigTag.type.create(c.arena, "c_longdouble"),
         .float80 => return ZigTag.type.create(c.arena, "f80"),
         .float128 => return ZigTag.type.create(c.arena, "f128"),
+        .@"enum" => @panic("TODO"),
+        .pointer,
+        .unspecified_variable_len_array,
+        .array,
+        .static_array,
+        .incomplete_array,
+        => @panic("TODO"),
         .func,
         .var_args_func,
         .old_style_func,
-        => return transFnType(c, scope, raw_ty, ty, source_loc, .{}),
+        => return transFnType(c, scope, ty, ty, source_loc, .{}),
+        .@"struct",
+        .@"union",
+        => {
+            var trans_scope = scope;
+            if (ty.isAnonymousRecord(c.comp)) {
+                const record_decl = ty.data.record;
+                const name_id = c.mapper.lookup(record_decl.name);
+                if (c.weak_global_names.contains(name_id)) trans_scope = &c.global_scope.base;
+            }
+            const name = c.decl_table.get(@intFromPtr(ty.data.record)).?;
+            return ZigTag.identifier.create(c.arena, name);
+        },
+        .attributed,
+        .typeof_type,
+        .typeof_expr,
+        => unreachable,
         else => return error.UnsupportedType,
     }
 }
 
-fn zigAlignment(bit_alignment: u29) u32 {
-    return bit_alignment / 8;
+/// Look ahead through the fields of the record to determine what the alignment of the record
+/// would be without any align/packed/etc. attributes. This helps us determine whether or not
+/// the fields with 0 offset need an `align` qualifier. Strictly speaking, we could just
+/// pedantically assign those fields the same alignment as the parent's pointer alignment,
+/// but this helps the generated code to be a little less verbose.
+fn headFieldAlignment(record_decl: *const Type.Record) ?c_uint {
+    const bits_per_byte = 8;
+    const parent_ptr_alignment_bits = record_decl.type_layout.pointer_alignment_bits;
+    const parent_ptr_alignment = parent_ptr_alignment_bits / bits_per_byte;
+    var max_field_alignment_bits: u64 = 0;
+    for (record_decl.fields) |field| {
+        if (field.ty.getRecord()) |field_record_decl| {
+            const child_record_alignment = field_record_decl.type_layout.field_alignment_bits;
+            if (child_record_alignment > max_field_alignment_bits)
+                max_field_alignment_bits = child_record_alignment;
+        } else {
+            const field_size = field.layout.size_bits;
+            if (field_size > max_field_alignment_bits)
+                max_field_alignment_bits = field_size;
+        }
+    }
+    if (max_field_alignment_bits != parent_ptr_alignment_bits) {
+        return parent_ptr_alignment;
+    } else {
+        return null;
+    }
+}
+
+/// This function returns a ?c_uint to match Clang's behaviour of using c_uint.
+/// This can be changed to a u29 after the Clang frontend for translate-c is removed.
+fn alignmentForField(
```

Comment:
It looks like the reason this function is complicated is because of differences in alignment in Zig and C but I'd like it to document that.